src/vanilla/store.ts

```diff
@@ -218,18 +221,88 @@ export const createStore = () => {
     return nextAtomState
   }
 
+  const setAtomValueOrPromise = <Value>(
+    atom: Atom<Value>,
+    valueOrPromise: Value,
+    nextDependencies?: NextDependencies,
+    abortPromise?: () => void
+  ): AtomState<Value> => {
+    if (valueOrPromise instanceof Promise) {
+      let continuePromise: (next: Promise<Awaited<Value>>) => void
+      const promise: Promise<Awaited<Value>> & PromiseMeta<Awaited<Value>> =
+        new Promise((resolve, reject) => {
+          let settled = false
+          valueOrPromise.then(
+            (v) => {
+              if (!settled) {
+                settled = true
+                const prevAtomState = getAtomState(atom)
+                // update dependencies, that could have changed
+                const nextAtomState = setAtomValue(
+                  atom,
+                  promise as Value,
+                  nextDependencies
+                )
+                resolvePromise(promise, v)
+                resolve(v)
+                if (prevAtomState?.d !== nextAtomState.d) {
+                  mountDependencies(atom, nextAtomState, prevAtomState?.d)
```

Comment:
Not super confident if this is the right solution, but this passes https://github.com/pmndrs/jotai/pull/1851/files#diff-48e03c388e7ed5d11021367c48a94ccc79febd8d1105af9922e7429279a04ed6R212 test.